pkg/reconciliation/reconcile_racks.go

```diff
@@ -1806,97 +1809,99 @@ func (rc *ReconciliationContext) startCassandra(endpointData httphelper.CassMeta
 	rc.ReqLogger.Info("reconcile_racks::startOneNodePerRack")
 
-	rackReadyCount := map[string]int{}
-	for _, rackInfo := range rc.desiredRackInformation {
-		rackReadyCount[rackInfo.RackName] = 0
-	}
+	labelSeedBeforeStart := readySeeds == 0 && !rc.hasAdditionalSeeds()
 
-	for _, pod := range rc.dcPods {
-		rackName := pod.Labels[api.RackLabel]
-		if isServerReady(pod) {
-			rackReadyCount[rackName]++
-		}
-	}
+	for _, statefulSet := range rc.statefulSets {
 
-	// if the DC has no ready seeds, label a pod as a seed before we start Cassandra on it
-	// and also consider additional seeds
+		pod := rc.getDCPodByName(statefulSet.Name + "-0")
+		notReady, err = rc.startNode(pod, labelSeedBeforeStart, endpointData)
+		if notReady {
+			break
+		}
 
-	externalSeedPoints := 0
-	if existingEndpoints, err := rc.GetAdditionalSeedEndpoint(); err == nil {
-		externalSeedPoints = len(existingEndpoints.Subsets[0].Addresses)
 	}
+	return
+}
+
+// startAllNodes starts all nodes in the datacenter in a deterministic order, dictated by the order
+// in which the racks were declared in the datacenter spec, and in such a way that at any given
+// time, the number of started nodes in a rack cannot be greater or lesser than the number of
+// started nodes in all other racks by more than 1. It returns true if a Cassandra node is not ready
+// yet, and an error if that node also failed to start.
+func (rc *ReconciliationContext) startAllNodes(endpointData httphelper.CassMetadataEndpoints) (notReady bool, err error) {
+	rc.ReqLogger.Info("reconcile_racks::startAllNodes")
 
-	labelSeedBeforeStart := readySeeds == 0 && len(rc.Datacenter.Spec.AdditionalSeeds) == 0 && externalSeedPoints == 0
+	for podRankWithinRack := 0; ; podRankWithinRack++ {
 
-	rackThatNeedsNode := ""
-	for rackName, readyCount := range rackReadyCount {
-		if readyCount > 0 {
-			continue
-		}
-		rackThatNeedsNode = rackName
-		for _, pod := range rc.dcPods {
-			mgmtApiUp := isMgmtApiRunning(pod)
-			if !isServerReadyToStart(pod) || !mgmtApiUp {
-				continue
-			}
-			podRack := pod.Labels[api.RackLabel]
-			if podRack == rackName {
-				// this is the one exception to all seed labelling happening in labelSeedPods()
-				if labelSeedBeforeStart {
-					patch := client.MergeFrom(pod.DeepCopy())
-					pod.Labels[api.SeedNodeLabel] = "true"
-					if err := rc.Client.Patch(rc.Ctx, pod, patch); err != nil {
-						return "", err
-					}
+		done := true
+		for _, statefulSet := range rc.statefulSets {
 
-					rc.Recorder.Eventf(rc.Datacenter, corev1.EventTypeNormal, events.LabeledPodAsSeed,
-						"Labeled pod a seed node %s", pod.Name)
+			maxPodRankInThisRack := int(*statefulSet.Spec.Replicas) - 1
+			if podRankWithinRack <= maxPodRankInThisRack {
 
-					// sleeping five seconds for DNS paranoia
-					time.Sleep(5 * time.Second)
-				}
-				if err := rc.startCassandra(endpointData, pod); err != nil {
-					return "", err
+				pod := rc.getDCPodByName(statefulSet.Name + "-" + strconv.Itoa(podRankWithinRack))
```

Comment:
getStatefulSetPodNameForIdx(sts, podRankWithinRack) would have probably worked fine, no need to replicate the naming structure here. Doesn't really matter though.

Reply:
Ah nice, I missed that function, sorry. Will use it.